source/Engine.cpp

```diff
@@ -593,6 +593,18 @@ void Engine::Step(bool isActive)
 		player.SetPlanet(flagship->GetPlanet());
 
 	const System *currentSystem = player.GetSystem();
+
+	// Check through the list of NPC ships from active missions
+	// to hit their "on encounter" triggers
+	// You can only encounter an NPC if you're not cloaked
+	if(flagship && currentSystem && flagship->Cloaking() == 0)
+		for(const Mission &mission : player.Missions())
+			for(const NPC &npc : mission.NPCs())
+				for(const shared_ptr<Ship> &npcShip : npc.Ships())
+					// You can only encounter an NPC if *they're* not cloaked
+					if(npcShip->GetSystem() == currentSystem && npcShip->Cloaking() == 0)
+						eventQueue.emplace_back(flagship, npcShip, ShipEvent::ENCOUNTER);
```

Comment:
also the way this looks, you keep encountering the npc after the first time?

Reply:
As I said, I recognize the _potential_ performance concern; however, _in practice_, it does not appear to cause any problems. If it is determined to be quite necessary to eliminate even the potential, I can of course rework it, but the only one of those actions that currently has a clear "hook" to hang this on is player entering the system, so it would be quite a bit of extra code for what is currently only a theoretical problem.

As for re-encountering, I was concerned, too, but all NPC actions are designed to trigger only once.

Reply:
Add: not the number of missions matters, the number of missions with npcs

Reply:
A fair request. I've done some more testing.

I picked up my fleet, accepted every Wanderer mission available there, then headed to Farpoint and accepted every human mission available there. This put me at over 30 active missions. I took off and landed a few more times, accepting all the Escort and Bounty missions, until I had a total of 18 of those. Then I took off again, flew around for a bit, let my fleet engage some pirates, and jumped over to Almaaz to fight more. I quit, checked out `master`, cleaned and ran that code, took off again, and did the same things.

On **both** branches, my CPU was averaging about 33% with spikes to 40%, and my GPU was averaging about 36%.

Again, this is on an M1 Max MacBook Pro, which has a _relatively_ powerful CPU and GPU, so it is possible that a less capable machine would show a difference in performance.